File: leetcode/problems/0001-two-sum/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> seen;  // giá trị -> chỉ số
        for (int i = 0; i < (int)nums.size(); ++i) {
            auto it = seen.find(target - nums[i]);
            if (it != seen.end()) return {it->second, i};
            seen[nums[i]] = i;  // ghi SAU khi tra, tránh ghép với chính nó
        }
        return {};  // đề bảo đảm luôn có đáp án
    }
};

// int main() {
//     Solution s;
//     vector<int> nums{2, 7, 11, 15};
//     auto r = s.twoSum(nums, 9);
//     cout << r[0] << ' ' << r[1] << '\n';  // 0 1
// }

File: leetcode/problems/0001-two-sum/solution.py
from typing import Dict, List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen: Dict[int, int] = {}  # giá trị -> chỉ số
        for i, x in enumerate(nums):
            if target - x in seen:
                return [seen[target - x], i]
            seen[x] = i  # ghi SAU khi tra
        return []


if __name__ == "__main__":
    print(Solution().twoSum([2, 7, 11, 15], 9))  # [0, 1]
